Renesas: Smartbond: Add MEMC Driver Support

ioannis-karachalios opened the pull request.

ioannis-karachalios committed:
drivers: memc: smartbond: Add support for the memory driver class.

Add support for the memory controller by utilizing QSPIC2. The latter is
capable to drive both NOR and PSRAM memory devices. For this to work,
the RAM driving mode is enabled.

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

ioannis-karachalios committed:
dts: renesas: smartbond: Add support for the memory driver class.

Update DTS and board configurations to support memory controller (QSPIC2).

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

ioannis-karachalios committed:
samples: drivers: memc: Add support for Smartbond Pro DevKit.

Add configuration and overlay files to support the DA1469x development kit.

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

ioannis-karachalios committed:
drivers: clock_control: smartbond: Add support for the memc driver.

1. Update the clock control driver so it can update timing settings for
   QSPIC2 following system clock transitions (translated based on
   AHB AMBA bus clock).
2. Remove the QSPIC related subroutines and use the respective HAL API
   which is now available.
3. Add support for PM (CONFIG_PM_DEVICE). This is required as QSPIC2
   register file is powered by PD_SYS which is turned off during device
   sleep and so registers contents are lost (in contrast to QSPIC which
   is used to drive the flash memory).

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

ioannis-karachalios committed:
tests: drivers: memc/ram: Add support for Smartbond Pro DevKit.

Add configuration and overlay files to support the DA1469x development kit.

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

ioannis-karachalios commented:
Kind reminder for someone to start looking at this PR. Thanks in advance.

ioannis-karachalios commented on drivers/memc/memc_smartbond_nor_psram.c:
```diff
@@ -0,0 +1,258 @@
+	/* Apply the min. required settings before performing any transaction in manual mode. */
+	qspic_ctrlmode_reg = QSPIC2->QSPIC2_CTRLMODE_REG;
+	QSPIC2_CTRLMODE_REG_SET_FIELD(QSPIC_CLK_MD, qspic_ctrlmode_reg,
+			DT_INST_PROP(0, clock_mode));
```
The reason I decided to make this a single-instance driver is because I cannot think of any use case that requires multiple instances; meaning that only a single memory device can be connected physically to the SoC.

blauret commented:
@andrzej-kaczmarek, I think it's waiting for your review.

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,254 @@
+    type: int
+    default: 0x5D
+    description: |
+      Device type, part of device ID, used for verification during MEMC initialization.
```
all default values in bindings require justification (https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values)

ditto for all the other properties with default values in this file

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,254 @@
+    type: int
+    default: 0xF5
+    description: |
+      Commnad to exit the QPI mode supported by a memory device
```
```suggestion
      Command to exit the QPI mode supported by a memory device
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,254 @@
+    default: 0xE040
+    description: |
+      Device density, part of device ID, used for verification during MEMC initialization.
+      [7:0] should reflect the density value ifself and [15:8] should reflect the mask
```
```suggestion
      [7:0] should reflect the density value itself and [15:8] should reflect the mask
```

ioannis-karachalios committed:
boards: renesas: dts: Add overlay files to support the MEMC

In order to avoid defining almost the same overlays in the available
sample codes, tests and user applications, a common overlay file
per memory type is demonstrated under the boards dts folder.
Currently only the PSRAM interface is supported and the APS6404L
PSRAM QSPI memory device is demonstrated. In doing so, an
application code will only have to define another overlay file explicitly,
under application's board folder, to overwrite the default QSPI
controller's settings. In either case, users should explicitly
invoke the requested overlay files  at 'west build' invokation
via the DTC_OVERLAY_FILE system variable.

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

ioannis-karachalios commented on drivers/clock_control/clock_control_smartbond.c:
```diff
@@ -370,12 +348,21 @@ int z_smartbond_select_sys_clk(enum smartbond_clock sys_clk)
+#if DT_PROP(DT_NODELABEL(memc), is_ram)
+		da1469x_qspi_set_tcem(SystemCoreClock, DT_PROP(DT_NODELABEL(memc), tcem_max_us));
+#endif
+#endif
```
Correct. Applying memory settings has been moved to a subroutine.

blauret commented:
@kartben, Could you kindly review again!!! I think we are there with this one.

blauret commented:
@andrzej-kaczmarek ping

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    type: int
+    required: true
+    description: |
+      Time in microsends (us) the memory device can accept the next command following a SW reset.
```
```suggestion
      Time in microseconds (us) the memory device can accept the next command following a SW reset.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+      access in case the total time exceeds the defined value
+      (at the cost of extra cycles required for re-sending the instruction,
+      address and dummy bytes, if any). This setting is meaningful only if
+      is-ram is present. This value reflects the max. time in microsends the
```
```suggestion
      is-ram is present. This value reflects the max. time in microseconds the
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    type: int
+    required: true
+    description: |
+      Manafacturer ID, part of device ID, used to verify the memory device used.
```
```suggestion
      Manufacturer ID, part of device ID, used to verify the memory device used.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    type: boolean
+    description: |
+      If present, the memory device will enter the QPI mode which typically reflects that
+      all bytes be sent in qaud bus mode. It's a pre-requisite that read and write
```
```suggestion
      all bytes be sent in quad bus mode. It's a pre-requisite that read and write
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    default: "addr-range-24bit"
+    description: |
+      Address size to use in auto mode. In 24-bit mode up to 16MB can be
+      accessed whilst in 32-bit mode up to 32MB can be accssed which is
```
```suggestion
      accessed whilst in 32-bit mode up to 32MB can be accessed which is
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    type: int
+    description: |
+      Clock divider for QSPIC2 controller. The clock path of
+      this block is always DIV1 which reflect the current
```
```suggestion
      this block is always DIV1 which reflects the current
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+      (at the cost of extra cycles required for re-sending the instruction,
+      address and dummy bytes, if any). This setting is meaningful only if
+      is-ram is present. This value reflects the max. time in microsends the
+      #CS line can be driven low in a write/read bust access (required for
```
```suggestion
      #CS line can be driven low in a write/read burst access (required for
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    description: |
+      If present the extra byte will be sent after the dummy bytes, if any.
+      This should be useful if 3 dummy bytes are required. In such a case,
+      dummy-bytes-size should be set to 2.
```
typo?

```suggestion
      dummy-bytes-count should be set to 2.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+      - "quad-spi"
+    default: "single-spi"
+    description: |
+      Describes the mode of SPI bus during the instruction phase for singe/burst
```
```suggestion
      Describes the mode of SPI bus during the instruction phase for single/burst
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    type: int
+    default: 0x03
+    description: |
+      Read command for single/burst read accesses in auto mode. This is the opcode for single mode
+      supported by all memory devices.
```
```suggestion
      Read command for single/burst read accesses in auto mode. Default value is the opcode for single mode
      which is supported by all memory devices.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    type: int
+    default: 0x02
+    description: |
+      Write command for single/burst write accesses in auto mode. This is the opcode for single mode
+      supported by all memory devices.
```
```suggestion
      Write command for single/burst write accesses in auto mode. Default value is the opcode for single mode
      which is supported by all memory devices.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    description: |
+      Address size to use in auto mode. In 24-bit mode up to 16MB can be
+      accessed whilst in 32-bit mode up to 32MB can be accssed which is
+      the max. address space supported by QSPICx. Select the 24-bit mode
+      which is supported by all memory devices.
```
```suggestion
      the max. address space supported by QSPICx. Default value is 24-bit mode
      which is supported by all memory devices.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    default: "single-spi"
+    description: |
+      Describes the mode of SPI bus during the address phase for single/burst
+      read accesses in auto mode. The single mode should be supported by all
+      memory devices.
```
```suggestion
      read accesses in auto mode. Default value is single mode which should be supported by all
      memory devices.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    default: "single-spi"
+    description: |
+      Describes the mode of SPI bus during the instruction phase for single/burst
+      read accesses in auto mode. The single mode should be supported by all
+      memory devices.
```
```suggestion
      read accesses in auto mode. Default value is single mode which should be supported by all
      memory devices.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    default: "single-spi"
+    description: |
+      Describes the mode of SPI bus during the data phase for single/burst
+      read accesses in auto mode. The single mode should be supported by all
+      memory devices.
```
```suggestion
      read accesses in auto mode. Default value is single mode which should be supported by all
      memory devices.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+      - "quad-spi"
+    default: "single-spi"
+    description: |
+      Describes the mode of SPI bus during the dummy bytes phase for single/burst
+      read accesses in auto mode. The single mode should be supported by all
+      memory devices.
```
```suggestion
      Describes the mode of SPI bus during the dummy bytes phase for single/burst
      read accesses in auto mode. Default value is single mode which should be supported by all
      memory devices.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+      - "dual-spi"
+      - "quad-spi"
+    description: |
+      Describes the mode of SPI bus during the extra byte phase for single/burst
+      read accesses in auto mode. The single mode should be supported by all
+      memory devices.
```
```suggestion
      Describes the mode of SPI bus during the extra byte phase for single/burst
      read accesses in auto mode. Default value is single mode which should be supported by all
      memory devices.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+      - "quad-spi"
+    default: "single-spi"
+    description: |
+      Describes the mode of SPI bus during the address phase for single/burst
+      write accesses in auto mode. The single mode should be supported by all
+      memory devices.
```
```suggestion
      Describes the mode of SPI bus during the address phase for single/burst
      write accesses in auto mode. Default value is single mode which should be supported by all
      memory devices.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+    default: "single-spi"
+    description: |
+      Describes the mode of SPI bus during the instruction phase for singe/burst
+      write accesses in auto mode. The single mode should be supported by all
+      memory devices.
```
```suggestion
      write accesses in auto mode. Default value is single mode which should be supported by all
      memory devices.
```

kartben commented on dts/bindings/memory-controllers/renesas,smartbond-nor-psram.yaml:
```diff
@@ -0,0 +1,260 @@
+      - "quad-spi"
+    default: "single-spi"
+    description: |
+      Describes the mode of SPI bus during the data phase for single/burst
+      write accesses in auto mode. The single mode should be supported by all
+      memory devices.
```
```suggestion
      Describes the mode of SPI bus during the data phase for single/burst
      write accesses in auto mode. Default value is single mode which should be supported by all
      memory devices.
```

ioannis-karachalios replied:
Fixed all typos
